internal/contour/snapshot.go

```diff
@@ -0,0 +1,130 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package contour
+
+import (
+	"fmt"
+	"math"
+
+	envoy_api_v2_core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	xds "github.com/envoyproxy/go-control-plane/pkg/cache/types"
+	"github.com/envoyproxy/go-control-plane/pkg/cache/v2"
+	"github.com/sirupsen/logrus"
+)
+
+// SnapshotHandler implements the xDS snapshot cache
+type SnapshotHandler struct {
+	// xDSResources is a local cache representing the current state
+	// of any xDS resource Contour is managing.
+	xDSResources map[xds.ResponseType][]xds.Resource
+
+	// SnapshotVersion holds the current version of the snapshot
+	snapshotVersion int64
+
+	// snapshotCache is a snapshot-based cache that maintains a single versioned
+	// snapshot of responses for xDS resources that Contour manages
+	snapshotCache cache.SnapshotCache
+
+	logrus.FieldLogger
+}
+
+// NewSnapshotHandler returns an instance of SnapShotHandler
+func NewSnapshotHandler(c cache.SnapshotCache, logger logrus.FieldLogger) *SnapshotHandler {
+	return &SnapshotHandler{
+		snapshotCache: c,
+		FieldLogger:   logger,
+		xDSResources:  make(map[xds.ResponseType][]xds.Resource, 5),
+	}
+}
+
+// ConstantHash is a specialized node ID hasher used to allow
+// any instance of Envoy to connect to Contour regardless of the
+// service-node flag configured on Envoy.
+type ConstantHash string
+
+func (c ConstantHash) ID(*envoy_api_v2_core.Node) string {
+	return string(c)
+}
+
+func (c ConstantHash) String() string {
+	return string(c)
+}
+
+var _ cache.NodeHash = ConstantHash("")
+var DefaultHash = ConstantHash("contour")
+
+// UpdateSnapshot is called when any cache changes and
+// Envoy should be updated with a new configuration.
+//
+// It does not take into account a specific cache changing.
+// When called, all xDS resources are updated with a new version.
+// Envoy sees this as a noop, but could be improved in future refactorings.
+func (s *SnapshotHandler) UpdateSnapshot(xDSResources map[xds.ResponseType][]xds.Resource) {
```

Comment:
I believe that since both the CacheHandler and the EndpointsTranslator could call this concurrently, we need some synchronization around it.